keras_nlp/models/gpt2/gpt2_causal_lm.py

```diff
@@ -324,47 +335,109 @@ def next(prompt, state, index):
 
     def generate(
         self,
-        prompt,
-        max_length,
+        inputs,
+        max_length=None,
+        batch_size=None,
+        return_tensors=False,
     ):
-        """Generate text.
+        """Generate text given prompt `inputs`.
+
+        This method generates text based on given `prompt`.The sampling used for
+        generation can be set in the `compile` method.
 
-        This method generates text based on given `prompt`. Generation will
-        continue until `max_length` is met, and all tokens generated after
-        `end_token` will be truncated. The sampling strategy can be set in
-        the `compile` method.
+        If `inputs` are a batched `tf.data.Dataset` or `batch_size` is provided,
+        outputs will be generated "batch-by-batch" and concatenated. Otherwise,
+        all inputs will be handled in a single generation pass.
+
+        If a `preprocessor` is attached to the model, `inputs` should be
+        strings and returned sequences will be strings. Otherwise, inputs should
+        be preprocessed before calling `generate()`, and returned sequences will
+        be token ids.
 
         Args:
-            prompt: a string, string Tensor or string RaggedTensor. The prompt
-                text for generation.
-            max_length: int. The max length of generated sequence.
+            inputs: a string `tf.Tensor`, a `tf.data.Dataset` of strings, a
+                python string or a list of python strings. If no `preprocessor`
+                is attached to the model, inputs should instead be a nested
+                `tf.Tensor` or `tf.data.Dataset` with keys `"token_ids"` and
+                `"padding_mask"`.
+            max_length: int. The max length of the generated sequence.
+            batch_size: int. Only pass if `inputs` is a `tf.Tensor`. If set,
+                `inputs` will be split into `batch_size` batches handled
+                successively.
+            return_tensors: bool. Only used if a `preprocessor` is attached to
+                the model. If `True`, `tf.Tensor` outputs will be returned. If
+                `False`, a python list of strings will be return.
+
+        Returns:
+            A string or string list if `preprocessor` is set, and a integer
+            tensor of token IDs if `preprocessor` is `None`.
         """
-        if self.preprocessor is None:
+        input_is_scalar = False
+
+        if isinstance(inputs, tf.data.Dataset) and batch_size is not None:
```

Comment:
Without the batching functionality we could also kill this

Reply:
Yes, if we remove the `batch_size` argument to `generate()`, we could only handle two cases. A dataset or a single batch. I am open to it!

Reply:
Removed `batch_size` for now.